### PR TITLE
Dynamo asserts FSDP wrapped modules use_orig_param

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89457
* #89802
* #89520
* __->__ #89523

- This is a strict requirement given the way dynamo+FSDP is implemented,
  but isn't convenient to assert.
- By plumbing use_orig_param field on all wrapped modules, we can
  do this assertion inside dynamo

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -336,6 +336,12 @@ def __init__(
             if submodule not in self._ignored_modules:
                 submodule._is_fsdp_managed_module = True
 
+                # Dynamo only supports FSDP with use_orig_params=True.
+                # This is hacky, but I could not think of another way to add an assertion to dynamo
+                # for this, since Dynamo skips all the FSDP code frames and thus can't inspect the
+                # FSDP module directly
+                submodule._fsdp_use_orig_params = use_orig_params
+
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:344]: This FSDP change looks fine to me!

Comment (wconstab) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:344]: ok. the rest of the change is pretty trivial. I think the biggest risk is that we are cluttering up the submodules this way.  the dynamo change is not a risk, and the unit test works.

**Review by awgu — APPROVED** [review]

**Pull request closed without merging**